service/src/main/java/bio/terra/tanagra/app/controller/CohortsV2ApiController.java

```diff
@@ -0,0 +1,204 @@
+
+import static bio.terra.tanagra.service.accesscontrol.Action.CREATE;
+import static bio.terra.tanagra.service.accesscontrol.Action.DELETE;
+import static bio.terra.tanagra.service.accesscontrol.Action.READ;
+import static bio.terra.tanagra.service.accesscontrol.Action.UPDATE;
+import static bio.terra.tanagra.service.accesscontrol.ResourceType.COHORT;
+
+import bio.terra.tanagra.generated.controller.CohortsV2Api;
+import bio.terra.tanagra.generated.model.ApiCohortCreateInfoV2;
+import bio.terra.tanagra.generated.model.ApiCohortListV2;
+import bio.terra.tanagra.generated.model.ApiCohortUpdateInfoV2;
+import bio.terra.tanagra.generated.model.ApiCohortV2;
+import bio.terra.tanagra.generated.model.ApiCriteriaGroupV2;
+import bio.terra.tanagra.generated.model.ApiCriteriaV2;
+import bio.terra.tanagra.query.filtervariable.BooleanAndOrFilterVariable.LogicalOperator;
+import bio.terra.tanagra.service.AccessControlService;
+import bio.terra.tanagra.service.CohortService;
+import bio.terra.tanagra.service.StudyService;
+import bio.terra.tanagra.service.UnderlaysService;
+import bio.terra.tanagra.service.accesscontrol.ResourceId;
+import bio.terra.tanagra.service.accesscontrol.ResourceIdCollection;
+import bio.terra.tanagra.service.artifact.Cohort;
+import bio.terra.tanagra.service.artifact.Criteria;
+import bio.terra.tanagra.service.artifact.CriteriaGroup;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+
+@Controller
+public class CohortsV2ApiController implements CohortsV2Api {
+  private final StudyService studyService;
+  private final CohortService cohortService;
+  private final UnderlaysService underlaysService;
+  private final AccessControlService accessControlService;
+
+  @Autowired
+  public CohortsV2ApiController(
+      StudyService studyService,
+      CohortService cohortService,
+      UnderlaysService underlaysService,
+      AccessControlService accessControlService) {
+    this.studyService = studyService;
+    this.cohortService = cohortService;
+    this.underlaysService = underlaysService;
+    this.accessControlService = accessControlService;
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortV2> createCohort(String studyId, ApiCohortCreateInfoV2 body) {
+    accessControlService.throwIfUnauthorized(null, CREATE, COHORT, new ResourceId(studyId));
+
+    // Make sure underlay name and study id are valid.
+    underlaysService.getUnderlay(body.getUnderlayName());
+    studyService.getStudy(studyId);
+
+    // Generate random 10-character alphanumeric strings for the new cohort and user-facing IDs.
+    String newCohortId = RandomStringUtils.randomAlphanumeric(10);
+    String newUserFacingCohortId = RandomStringUtils.randomAlphanumeric(10);
```

Comment:
Why do cohorts have user facing IDs, since cohort IDs are generated by the backend? If it's for versioning, it would be clearer to name these `cohortId` and `revisionId` or something similar.

Reply:
Yes, it's for versioning. Cohort = filter on primary entity. User-facing cohort = group of one or more cohort revisions.

`cohortId` maps to a single row of the `cohort` table (= a single cohort revision), so I think that's clearer as is on the backend at least. Maybe renaming `userFacingCohortId` would be helpful? Suggestions welcome (`cohortRevisionGroup` was another name I considered).

Reply:
Even in your response you're describing them as "single cohort revisions". Perhaps `cohortId` and `cohortRevsionId`?

Reply:
`cohortId` is the primary key of the `cohort` table. To the backend, a cohort is just a filter on the primary entity. So one `cohortId` corresponds to one `Cohort` object. That's what I'm trying to preserve for clarity on the backend.

To the API, `cohortId` refers to the user-facing id, so this naming doesn't extend to callers of the API. It's just for the DB interactions and internal objects on the backend.

Reply:
Per offline conversation, I made a couple changes related to this thread:
- Updated the API to define a separate `CohortRevision` object. The `Cohort` object contains a reference to the latest revision, and the `Review` object contains a reference to the frozen revision at the time the review was created.
- Replaced `user_facing_cohort_id` with `cohort_revision_group_id` in the backend table and class. There's still only a single table on the backend for cohort revisions because there's only the `underlayName` that applies across revisions (and even that could change in the future depending on how we handle underlying data upgrades). But if we add more fields to the set of cohort revisions in the future, it would make sense to split this off into a separate table.
- Replaced `getCohortUserFacing` and similarly named methods with `getCohortLatestVersion` to make the backend code easier to follow and map to the API calls.

I think there's still some ambiguity when we talk about a Tanagra cohort, because sometimes it refers to the latest revision in a collection of revisions, and sometimes it refers to a frozen revision. Will bring this up for discussion with the PMs regarding whether/how to make this clearer in the UI.